synchronization/trigger_test.go

```diff
@@ -26,120 +26,133 @@ func newNopLogger() *nopLogger {
+func NewTestFlag(t testing.TB) *TestFlag {
+	return &TestFlag{
+		t:     t,
+		count: 0,
 	}
-	tickTime := time.Microsecond
-	p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", tickTime, logger, trigger, nil)
+}
 
-	requireAtomicGreaterThan(t, &handlerInvocationCount, 0, "expected trigger to have ticked once promptly")
+func (c *TestFlag) Toggle() {
+	atomic.StoreUint32(&(c.count), 1)
+}
+func (c *TestFlag) EventuallyToggled(message string) {
+	require.Eventually(c.t, func() bool { return atomic.CompareAndSwapUint32(&(c.count), 1, 0) }, time.Second, time.Millisecond, message)
+}
 
-	p.Stop()
+func (c *TestFlag) NotEventuallyToggled(message string) {
+	time.Sleep(time.Millisecond * 500)
+	require.True(c.t, atomic.LoadUint32(&(c.count)) == 0, message)
+}
+
+func TestPeriodicalTriggerStartsOk(t *testing.T) {
+	logger := newNopLogger()
+	triggerFlag := NewTestFlag(t)
+	ticker := synchronization.NewTimeTicker(time.Millisecond)
+	p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", ticker, logger, func() { triggerFlag.Toggle() }, nil)
+	defer p.Stop()
+	triggerFlag.EventuallyToggled("expected trigger to have ticked once immediately")
 }
 
-func TestPeriodicalTrigger_Stop(t *testing.T) {
+func TestPeriodicalTriggerNoTicksNoTriggers(t *testing.T) {
 	with.Concurrency(t, func(ctx context.Context, harness *with.ConcurrencyHarness) {
-		x := 0
-		p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", time.Millisecond*2, harness.Logger, func() { x++ }, nil)
+		ticker := synchronization.NewHookTicker() // does not tick by default
+		triggerFlag := NewTestFlag(t)
+		p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", ticker, harness.Logger, func() { triggerFlag.Toggle() }, nil)
+		defer p.Stop()
 		harness.Supervise(p)
-		p.Stop()
-		time.Sleep(3 * time.Millisecond)
-		require.Equal(t, 0, x, "expected no ticks")
+		triggerFlag.NotEventuallyToggled("expected no triggers when no ticks")
 	})
 }
 
-func TestPeriodicalTrigger_StopAfterTrigger(t *testing.T) {
+func TestPeriodicalTriggerTriggersOnTicks(t *testing.T) {
 	with.Concurrency(t, func(ctx context.Context, harness *with.ConcurrencyHarness) {
-		x := 0
-		p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", time.Millisecond, harness.Logger, func() { x++ }, nil)
+		ticker := synchronization.NewHookTicker()
+		ch := make(chan time.Time)
+		ticker.C_ = func() <-chan time.Time { return ch }
+		triggerFlag := NewTestFlag(t)
+		p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", ticker, harness.Logger, func() { triggerFlag.Toggle() }, nil)
+		defer p.Stop()
 		harness.Supervise(p)
-		time.Sleep(time.Microsecond * 1100)
-		p.Stop()
-		xValueOnStop := x
-		time.Sleep(time.Millisecond * 5)
-		require.Equal(t, xValueOnStop, x, "expected one tick due to stop")
-
+		for i := 0; i < 5; i++ {
+			ch <- time.Now()
```

Comment:
following @electricmonk's comment above, can this be wrapped in a `HookTicker` method (`ticker.tick()`)?